File: packages/sdk-java/src/main/java/io/settler/clients/JobsClient.java
package io.settler.clients;

import com.fasterxml.jackson.databind.JsonNode;
import io.settler.exceptions.*;
import io.settler.http.HttpExecutor;

import java.util.*;

/**
 * Client for reconciliation job operations.
 *
 * <pre>{@code
 * SettlerClient client = SettlerClient.create("sk_test");
 * JsonNode job = client.jobs().create("stripe", "2025-01-01T00:00:00Z", "2025-01-31T23:59:59Z", null);
 * JsonNode allJobs = client.jobs().list(1, 100, null, null);
 * }</pre>
 */
public class JobsClient {
    private final HttpExecutor http;

    /**
     * Creates a JobsClient with a shared HttpExecutor.
     *
     * @param http the shared HTTP executor
     */
    public JobsClient(HttpExecutor http) {
        this.http = http;
    }

    /**
     * @deprecated Use {@link #JobsClient(HttpExecutor)} via SettlerClient instead.
     */
    @Deprecated
    public JobsClient(String baseUrl, String apiKey) {
        this(new HttpExecutor(
            io.settler.SettlerConfig.builder().apiKey(apiKey).baseUrl(baseUrl).build()
        ));
    }

    /**
     * Create a new reconciliation job.
     *
     * @param provider  Payment provider (stripe, paypal, square, bank)
     * @param startDate Start of date range (ISO 8601)
     * @param endDate   End of date range (ISO 8601)
     * @param options   Optional job options (autoReconcile, notifyOnComplete)
     * @return The created job data
     * @throws SettlerException if the request fails
     */
    public JsonNode create(String provider, String startDate, String endDate, Map<String, Object> options) throws SettlerException {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("provider", provider);
        Map<String, String> dateRange = new LinkedHashMap<>();
        dateRange.put("start", startDate);
        dateRange.put("end", endDate);
        body.put("dateRange", dateRange);
        if (options != null) {
            body.put("options", options);
        }
        return http.post("/jobs", body);
    }

    /**
     * List reconciliation jobs with optional filtering.
     *
     * @param page     page number (nullable)
     * @param limit    items per page (nullable)
     * @param status   filter by status (nullable)
     * @param provider filter by provider (nullable)
     * @return paginated job list
     * @throws SettlerException if the request fails
     */
    public JsonNode list(Integer page, Integer limit, String status, String provider) throws SettlerException {
        Map<String, String> params = new LinkedHashMap<>();
        if (page != null) params.put("page", page.toString());
        if (limit != null) params.put("limit", limit.toString());
        if (status != null) params.put("status", status);
        if (provider != null) params.put("provider", provider);
        return http.get("/jobs", params);
    }

    /**
     * Get a reconciliation job by ID.
     *
     * @param jobId the job ID
     * @return the job details
     * @throws SettlerException if the request fails
     */
    public JsonNode get(String jobId) throws SettlerException {
        return http.get("/jobs/" + jobId, null);
    }

    /**
     * Run a reconciliation job.
     *
     * @param jobId the job ID
     * @return the run result
     * @throws SettlerException if the request fails
     */
    public JsonNode run(String jobId) throws SettlerException {
        return http.post("/jobs/" + jobId + "/run");
    }

    /**
     * Delete a reconciliation job.
     *
     * @param jobId the job ID
     * @throws SettlerException if the request fails
     */
    public void delete(String jobId) throws SettlerException {
        http.delete("/jobs/" + jobId);
    }
}

File: packages/sdk-csharp/Clients/JobsClient.cs
using Settler.Sdk.Models;

namespace Settler.Sdk;

/// <summary>Client for reconciliation job operations.</summary>
public sealed class JobsClient
{
    private readonly SettlerClient _client;

    internal JobsClient(SettlerClient client) => _client = client;

    /// <summary>Create a new reconciliation job.</summary>
    public async Task<DataResponse<Job>> CreateAsync(CreateJobRequest request, CancellationToken cancellationToken = default)
    {
        return await _client.RequestAsync<DataResponse<Job>>(
            HttpMethod.Post, "jobs", body: request, cancellationToken: cancellationToken);
    }

    /// <summary>List reconciliation jobs.</summary>
    public async Task<PaginatedResponse<Job>> ListAsync(
        int? page = null, int? limit = null, string? status = null, string? provider = null,
        CancellationToken cancellationToken = default)
    {
        var query = new Dictionary<string, string>();
        if (page.HasValue) query["page"] = page.Value.ToString();
        if (limit.HasValue) query["limit"] = limit.Value.ToString();
        if (status != null) query["status"] = status;
        if (provider != null) query["provider"] = provider;

        return await _client.RequestAsync<PaginatedResponse<Job>>(
            HttpMethod.Get, "jobs", query: query, cancellationToken: cancellationToken);
    }

    /// <summary>Get a job by ID.</summary>
    public async Task<DataResponse<Job>> GetAsync(string id, CancellationToken cancellationToken = default)
    {
        return await _client.RequestAsync<DataResponse<Job>>(
            HttpMethod.Get, $"jobs/{Uri.EscapeDataString(id)}", cancellationToken: cancellationToken);
    }

    /// <summary>Run a reconciliation job.</summary>
    public async Task<DataResponse<Job>> RunAsync(string id, CancellationToken cancellationToken = default)
    {
        return await _client.RequestAsync<DataResponse<Job>>(
            HttpMethod.Post, $"jobs/{Uri.EscapeDataString(id)}/run", cancellationToken: cancellationToken);
    }

    /// <summary>Delete a reconciliation job.</summary>
    public async Task DeleteAsync(string id, CancellationToken cancellationToken = default)
    {
        await _client.RequestAsync(
            HttpMethod.Delete, $"jobs/{Uri.EscapeDataString(id)}", cancellationToken: cancellationToken);
    }
}
